src/clj/com/climate/claypoole/lazy.clj

```diff
@@ -0,0 +1,249 @@
+;; The Climate Corporation licenses this file to you under under the Apache
+;; License, Version 2.0 (the "License"); you may not use this file except in
+;; compliance with the License.  You may obtain a copy of the License at
+;;
+;;   http://www.apache.org/licenses/LICENSE-2.0
+;;
+;; See the NOTICE file distributed with this work for additional information
+;; regarding copyright ownership.  Unless required by applicable law or agreed
+;; to in writing, software distributed under the License is distributed on an
+;; "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+;; or implied.  See the License for the specific language governing permissions
+;; and limitations under the License.
+
+(ns com.climate.claypoole.lazy
+  "Lazy threadpool tools for Clojure.
+
+  This namespace provides lazy versions of the parallel functions from
+  com.climate.claypoole. They will only run the outputs you realize, plus a few
+  more (a buffer), just like core pmap. The buffer is used to help keep the
+  threadpool busy.
+
+  Each parallel function also comes with a -buffer variant that allows you to
+  specify the buffer size. The non -buffer forms use the threadpool size as
+  their buffer size.
+
+  To use the threadpool most efficiently with these lazy functions, prefer the
+  unordered versions (e.g. upmap), since the ordered ones may starve the
+  threadpool of work. For instance, this pmap will take 6 milliseconds to run:
+    (lazy/pmap 2 #(Thread/sleep %) [4 3 2 1])
+  That's because it will not realize the 2 task until the 4 task is complete,
+  so one thread in the pool will sit idle for 1 millisecond."
+  (:refer-clojure :exclude [future future-call pcalls pmap pvalues])
+  (:require [com.climate.claypoole :as cp]
+            [com.climate.claypoole.impl :as impl])
+  (:import [java.util.concurrent LinkedBlockingQueue]))
+
+
+(defn- forceahead
+  "A sequence of s with the next buffer-size elements of it forced."
+  [buffer-size s]
+  (map (fn [x _] x)
+       s
+       (concat (drop buffer-size s) (repeat nil))))
+
+(defn- pool-closer
+  "Given a sequence, concat onto its end a lazy sequence that will shutdown a
```

Comment:
i don't think the docstring is accurate. Or I don't understand it. This is just calling shutdown after the last element right? Maybe a nicer interface is to get a no-arg function defaulted to identity if nil. That way you have only 2 args and it's more reusable (doesn't depend on shutdown), not that you are going to reuse it, but it would be easier to understand.